flink/src/main/java/org/apache/iceberg/flink/data/FlinkValueWriters.java

```diff
@@ -0,0 +1,240 @@
+  }
+
+  static ValueWriter<TimestampData> timestampMicros() {
+    return TimestampMicrosWriter.INSTANCE;
+  }
+
+  static ValueWriter<DecimalData> decimal(int precision, int scale) {
+    return new DecimalWriter(precision, scale);
+  }
+
+  static <T> ValueWriter<ArrayData> array(ValueWriter<T> elementWriter, LogicalType elementType) {
+    return new ArrayWriter<>(elementWriter, elementType);
+  }
+
+  static <K, V> ValueWriter<MapData> arrayMap(ValueWriter<K> keyWriter, LogicalType keyType,
+                                              ValueWriter<V> valueWriter, LogicalType valueType) {
+    return new ArrayMapWriter<>(keyWriter, keyType, valueWriter, valueType);
+  }
+
+  static <K, V> ValueWriter<MapData> map(ValueWriter<K> keyWriter, LogicalType keyType,
+                                         ValueWriter<V> valueWriter, LogicalType valueType) {
+    return new MapWriter<>(keyWriter, keyType, valueWriter, valueType);
+  }
+
+  static ValueWriter<RowData> row(List<ValueWriter<?>> writers, List<LogicalType> types) {
+    return new RowWriter(writers, types);
+  }
+
+  private static class StringWriter implements ValueWriter<StringData> {
+    private static final StringWriter INSTANCE = new StringWriter();
+
+    private StringWriter() {
+    }
+
+    @Override
+    public void write(StringData s, Encoder encoder) throws IOException {
+      // toBytes is cheaper than Avro calling toString, which incurs encoding costs
+      encoder.writeString(new Utf8(s.toBytes()));
+    }
+  }
+
+  private static class DecimalWriter implements ValueWriter<DecimalData> {
+    private final int precision;
+    private final int scale;
+    private final ThreadLocal<byte[]> bytes;
+
+    private DecimalWriter(int precision, int scale) {
+      this.precision = precision;
+      this.scale = scale;
+      this.bytes = ThreadLocal.withInitial(() -> new byte[TypeUtil.decimalRequiredBytes(precision)]);
+    }
+
+    @Override
+    public void write(DecimalData d, Encoder encoder) throws IOException {
+      encoder.writeFixed(DecimalUtil.toReusedFixLengthBytes(precision, scale, d.toBigDecimal(), bytes.get()));
+    }
+  }
+
+  private static class TimeMicrosWriter implements ValueWriter<Integer> {
+    private static final TimeMicrosWriter INSTANCE = new TimeMicrosWriter();
+
+    @Override
+    public void write(Integer timeMills, Encoder encoder) throws IOException {
+      encoder.writeLong(timeMills * 1000);
+    }
+  }
+
+  private static class TimestampMicrosWriter implements ValueWriter<TimestampData> {
+    private static final TimestampMicrosWriter INSTANCE = new TimestampMicrosWriter();
+
+    @Override
+    public void write(TimestampData timestampData, Encoder encoder) throws IOException {
+      long micros = timestampData.getMillisecond() * 1000 + timestampData.getNanoOfMillisecond() / 1000;
+      encoder.writeLong(micros);
+    }
+  }
+
+  private static class ArrayWriter<T> implements ValueWriter<ArrayData> {
```

Comment:
Eventually, we should refactor this into a base class for array data, so that the encoder parts are shared between Flink and Spark. Not something we should do right now, though.